hexa/plugins/connector_gcs/models.py

```diff
@@ -0,0 +1,174 @@
+import typing
+from logging import getLogger
+
+from django.contrib.auth.models import AnonymousUser
+from django.core.exceptions import ValidationError
+from django.db import models
+from django.db.models import Q
+from django.template.defaultfilters import pluralize
+from django.urls import reverse
+from django.utils.translation import gettext_lazy as _
+
+from hexa.catalog.models import Datasource
+from hexa.core.models import Base
+from hexa.core.models.base import BaseQuerySet
+from hexa.core.models.cryptography import EncryptedTextField
+from hexa.user_management import models as user_management_models
+from hexa.user_management.models import Permission, PermissionMode
+
+logger = getLogger(__name__)
+
+
+class Credentials(Base):
+    """We actually only need one set of credentials. These "principal" credentials will be then used to generate
+    short-lived credentials with a tailored policy giving access only to the buckets that the user team can
+    access"""
+
+    class Meta:
+        verbose_name = "GCS Credentials"
+        verbose_name_plural = "GCS Credentials"
+        ordering = ("service_account",)
+
+    service_account = models.CharField(max_length=30)
+    project_id = models.CharField(max_length=30)
+    client_id = models.CharField(max_length=30)
+    client_email = models.CharField(max_length=100)
+    client_x509_cert_url = models.CharField(max_length=150)
+    auth_uri = models.CharField(
+        max_length=100, default="https://accounts.google.com/o/oauth2/auth"
+    )
+    token_uri = models.CharField(
+        max_length=100, default="https://oauth2.googleapis.com/token"
+    )
+    auth_provider_x509_cert_url = models.CharField(
+        max_length=100, default="https://www.googleapis.com/oauth2/v1/certs"
+    )
+
+    private_key_id = EncryptedTextField(max_length=50)
+    private_key = EncryptedTextField()
+
+    @property
+    def display_name(self):
+        return self.service_account
+
+
+class BucketQuerySet(BaseQuerySet):
+    def filter_for_user(
+        self,
+        user: typing.Union[AnonymousUser, user_management_models.User],
+        mode: PermissionMode = None,
+        mode__in: typing.Sequence[PermissionMode] = None,
+    ):
+        if mode is not None and mode__in is not None:
+            raise ValueError('Please provide either "mode" or "mode_in" - not both')
+
+        if user.is_superuser or user.has_feature_flag("gcs"):
```

Comment:
If we don't filter the buckets at all, I would prefer if we dropped the `has_feature_flag` part and restrict the GCS plugin to super admins.